tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc

```
@@ -1401,6 +1401,23 @@ TEST_F(EnergyPlusFixture, HVACMultiSpeedHeatPump_ReportVariableInitTest)
     EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPump(2).CompPartLoadRatio, 0.1530992, 0.0001);
 
     state->dataGlobal->DoCoilDirectSolutions = false;
+
+    state->dataHVACMultiSpdHP->MSHeatPump(2).EMSOverrideCoilSpeedNumOn = true;
+    state->dataHVACMultiSpdHP->MSHeatPump(2).EMSOverrideCoilSpeedNumValue = 2;
+    QZnReq = 50000.00;
+    SimMSHP(*state, MSHeatPumpNum, FirstHVACIteration, AirLoopNum, QSensUnitOut, QZnReq, OnOffAirFlowRatio);
+    EXPECT_EQ(state->dataHVACMultiSpdHP->MSHeatPumpReport(2).SpeedNum, 2);
+    EXPECT_EQ(state->dataHVACMultiSpdHP->MSHeatPumpReport(2).CycRatio, 1.0);
+    EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPumpReport(2).SpeedRatio, 0.4686, 0.0001);
+
+    state->dataHVACMultiSpdHP->MSHeatPump(2).EMSOverrideCoilSpeedNumValue = 1;
+    QZnReq = 10000.00;
+    SimMSHP(*state, MSHeatPumpNum, FirstHVACIteration, AirLoopNum, QSensUnitOut, QZnReq, OnOffAirFlowRatio);
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).Temp, 26.546664, 0.0001);
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).HumRat, 0.008, 0.0001);
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).Enthalpy, 47077.4613, 0.0001);
+    EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPump(2).CompPartLoadRatio, 0.1530992, 0.0001);
+
```

I tested the unit test. It worked as expected. When I removed EMS overrides (lines 1405, 1406 and 1413), the unit test still passes. That means the override values are the same as calculated one. In other words, the EMS override does not override calculated values.

Then I set up value = 1 on line 1406 and value = 2 at line 1413. Then speedNum = 1 and CycRatio = 1 in the first case, as expected. However, the second case provide a fatal message as "DX unit compressor speed calculation failed: speed limits exceeded...." In other words, when the override value is greater than the calculated value, no solution will be provided.

Here is my suggestion.

1. Allow the program to calculate SpeedNum also.
2. Compare EMS override value with SpeedNum
3. If EMS > SpeedNum, use calculated SppedNum, and provide warning: Override Value will not be used, Simulation continues...
4. If EMS < = SpeedNum, use EMS value

---

1. The test case does successfully override the calculations, as when `EMSOverrideCoilSpeedNumOn = true;` in line 1406, the program used the [EMS control branch](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L310) instead of the original branch.

2. When removing lines 1405 and 1406, the unit test still passed, because without EMS override, the program redirection to normal control [at here](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L358), and calculated speed num under `QZnReq = 50000.00;` is also 2, same as the EMS override case. And under the same speed level, the `CycRatio` and `SpeedRatio` are supposed be the same.

---

3. [Line 1413 to 1419 ](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc#L1419) tests that the EMS override logic reproduces the same system partial load output with the normal control. With SpeedNum set back to 1 by EMS override value, the partial load ratio is the same  with the normal control calculated [partial speed load ratio](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc#L1401).

---

4. I'm not sure about the logic that if the EMS override SpeedNum is not solvable, we switch back to the normal control. On the one hand, the user-defined EMS program should take full control and should be responsible for making sense of the control logic. If the EMS override value is not solvable, it should report an error. On the other hand, I also have the feeling that it's hard for an external program to achieved a solvable logic with fewer handlers than the defined control program in E+, so it might be fairly easy for the control to reach an unsolvable state. @Myoldmopar @mjwitte Edwin and Mike, were there similar issues in Plant Loop override control as well? How was the problem solved there?

---

Here is my thought about Item 4.

If we know that the EMS override is not solvable, why do we need override?

---

Because I feel like that's what the user requests, to change the logic. If the request is not legitimate, then they should be notified to modify their desired logic. I think users would only use this feature if they want to override the Non-EMS logic, and overriding their overrides would be a bit confusing.

---

I would expect EMS override to be done using short timesteps, with logic based on temperature differences, eg. if zone temp is off by 1.0C use speed 1, if off by 2.0C use speed 2.  The part load with in a speed level is not needed. Just always use part load of 1.0 with EMs override of speed number and run. nothing to solve.  Control will be realistic but different than usual energyplus load-based, long timestep approach.